File: shared/validation.py
"""
Shared validation utilities
Used across microservices
"""
import re


def is_valid_email(email: str) -> bool:
    """
    Validate email format
    
    Args:
        email: Email string to validate
        
    Returns:
        True if valid email format, False otherwise
    """
    if not email or not isinstance(email, str):
        return False
    
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def sanitize_string(input_str: str, max_length: int = 255) -> str:
    """
    Sanitize user input string
    
    Args:
        input_str: Input string to sanitize
        max_length: Maximum allowed length
        
    Returns:
        Sanitized string
    """
    if not input_str:
        return ""
    
    # Remove potentially dangerous characters
    sanitized = re.sub(r'[<>\"\'&]', '', str(input_str))
    
    # Trim to max length
    return sanitized[:max_length].strip()


def validate_positive_number(value, field_name="value"):
    """
    Validate that a value is a positive number
    
    Args:
        value: Value to validate
        field_name: Name of field for error message
        
    Returns:
        Validated number
        
    Raises:
        ValueError: If value is not a positive number
    """
    try:
        num = float(value)
        if num < 0:
            raise ValueError(f"{field_name} must be positive")
        return num
    except (TypeError, ValueError) as e:
        raise ValueError(f"{field_name} must be a valid positive number: {e}")

File: shared/validation.js
/**
 * Shared validation utilities for Node.js services
 */

/**
 * Validate email format
 * @param {string} email - Email to validate
 * @returns {boolean} True if valid email format
 */
function isValidEmail(email) {
    if (!email || typeof email !== 'string') {
        return false;
    }
    
    const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return pattern.test(email);
}

/**
 * Sanitize user input string
 * @param {string} inputStr - Input string to sanitize
 * @param {number} maxLength - Maximum allowed length
 * @returns {string} Sanitized string
 */
function sanitizeString(inputStr, maxLength = 255) {
    if (!inputStr) {
        return '';
    }
    
    // Remove potentially dangerous characters
    const sanitized = String(inputStr).replace(/[<>"'&]/g, '');
    
    // Trim to max length
    return sanitized.substring(0, maxLength).trim();
}

/**
 * Validate that a value is a positive number
 * @param {any} value - Value to validate
 * @param {string} fieldName - Name of field for error message
 * @returns {number} Validated number
 * @throws {Error} If value is not a positive number
 */
function validatePositiveNumber(value, fieldName = 'value') {
    const num = parseFloat(value);
    
    if (isNaN(num)) {
        throw new Error(`${fieldName} must be a valid number`);
    }
    
    if (num < 0) {
        throw new Error(`${fieldName} must be positive`);
    }
    
    return num;
}

/**
 * Validate pagination parameters
 * @param {number} offset - Offset value
 * @param {number} limit - Limit value
 * @returns {object} Validated pagination object
 */
function validatePagination(offset = 0, limit = 10) {
    const validOffset = Math.max(0, parseInt(offset) || 0);
    const validLimit = Math.min(100, Math.max(1, parseInt(limit) || 10));
    
    return { offset: validOffset, limit: validLimit };
}

module.exports = {
    isValidEmail,
    sanitizeString,
    validatePositiveNumber,
    validatePagination
};
